surefire-providers/surefire-junit-platform/src/main/java/org/apache/maven/surefire/junitplatform/JUnitPlatformProvider.java

```diff
@@ -187,12 +188,12 @@ private LauncherDiscoveryRequest buildLauncherDiscoveryRequestForRerunFailures(
         LauncherDiscoveryRequestBuilder builder = request().filters( filters ).configurationParameters(
                 configurationParameters );
         // Iterate over recorded failures
-        for ( TestIdentifier identifier : adapter.getFailures().keySet() )
+        for ( TestIdentifier identifier : new HashSet<>( adapter.getFailures().keySet() ) )
         {
             // Extract quantified test name data
             String[] classMethodName = adapter.toClassMethodNameWithoutPlan( identifier );
-            String className = classMethodName[1];
-            String methodName = classMethodName[3];
+            String className = classMethodName[0];
```

Comment:
It would be worth to use OOP and not string array because now I can see how it is difficult to understand the indexes of the array and their meaning. So we have a couple of clsaa and method, and one group is the traditional and second is named. So the class can have annotation as well.

Reply:
I agree, but I don't think that's within the scope of this issue _(A simple bug fix)_.

Reply:
yes, right, we can do it after this PR. And we do not need to have Jira issue for a small internal code improvement.